### PR TITLE
build: automate syntax functions definitions

### PR DESCRIPTION
**Pull request by l-zeuch**
**Please describe the changes this pull request does and why it should be merged:**

Automate the syntax function definitions using lytfs.

In the meantime, also bump actions/checkout to v4.

Signed-off-by: Luca Zeuch <l-zeuch@email.de>

**Terms**
- [x] I agree to follow this project's [Code of Conduct](CODE_OF_CONDUCT.md)
- [x] I have read and understood this project's [Contributing Guidelines](../CONTRIBUTING.md)

### CODE REVIEW

File: scripts/gen_funcs.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+import os
+
+def gen_code(file):
```
Comment (jo3-l) [inline scripts/gen_funcs.py:20]:
```suggestion
def gen_code(funcs):
```
and change `gen_code` to take a list of functions? (So the caller should read the `lytfs` file and split lines.)

File: scripts/gen_funcs.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    with open(file, "rt") as fin:
+        code = ""
+        for line in fin:
+            line = line.strip()
+            code += f"syn keyword yagFunc {line} contained\n"
+
+    return code
+
+def write_file(code):
+    outfile = "syntax/yagpdbcc/functions.vim"
```
Comment (jo3-l) [inline scripts/gen_funcs.py:30]: This should be a parameter to the function.

File: scripts/gen_funcs.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    if "syntax" not in os.listdir():
+        try:
+            os.mkdir("syntax")
+        except OSError as error:
+            print(error)
+            os._exit(1)
+
+    # clear the old file
+    with open(outfile, "wt") as fout:
+        fout.write("")
```
Comment (jo3-l) [inline scripts/gen_funcs.py:41]: `open(outfile, "w+")` will create if needed and truncate.

File: scripts/gen_funcs.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    # insert the header comment and append the code
+    with open("scripts/comment.txt", "rt") as fin, open(outfile, "wt") as fout:
+        for line in fin:
+            fout.write(line)
```
Comment (jo3-l) [inline scripts/gen_funcs.py:46]: fout.write(fin.read())

**Review by jo3-l — COMMENTED** [review]
Some suggestions to make your script more Pythonic.

**Comment by l-zeuch** [comment]
Joe, I think I understood you correctly...

Your feedback is addressed in [954d6fa](https://github.com/l-zeuch/yagpdb.vim/pull/83/commits/954d6fa9fc2c98e1bf21c59c44a2c401185c2b00).

**Review by LRitzdorf — APPROVED** [review]
Looks neat! I'll have to check on exactly what/how lytfs does, but the changes here make sense.

**Comment by LRitzdorf** [comment]
Oh, cool! Nice work, @jo3-l

File: scripts/gen_funcs.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+import os
+
+def gen_code(funcs):
+    code = ""
+    for func in funcs:
+        code += f"syn keyword yagFunc {func} contained\n"
```
[the remark this answers is not in the document]
Comment (l-zeuch) [reply scripts/gen_funcs.py:23]: That'd change the return type of `str` to `Iterable`, no?
So not really viable...

File: scripts/gen_funcs.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+
+def main():
+    file = 'syntax/funcs'
+
+    funcs = []
+    with open(file, "rt") as fin:
+        for line in fin:
+            line = line.strip()
+            funcs.append(line)
```
[the remark this answers is not in the document]
Comment (l-zeuch) [reply scripts/gen_funcs.py:49]: Makes sense here, a bit more readable that way.

File: scripts/comment.txt
Comment (l-zeuch) [inline scripts/comment.txt:1]: This should perhaps be more aptly named `funcs_comment.vim`.

What do you think?

File: scripts/gen_funcs.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (LRitzdorf) [reply scripts/gen_funcs.py:23]: Not if we `'\n'.join([...])`ed it. Up to you, though

[the remark this answers is not in the document]
Comment (l-zeuch) [reply scripts/gen_funcs.py:23]: True; will do just that.

File: scripts/comment.txt
Comment (LRitzdorf) [reply scripts/comment.txt:1]: I'm not really sure what, if any, naming standards there are for "header comment" files like these. Personally, I think either one makes sense

Comment (l-zeuch) [reply scripts/comment.txt:1]: I'll just leave this as-is, it's not like this name hurts anyone.

**Comment by l-zeuch** [comment]
I'm still a bit awkward about merging this — I'll probably wait until
Wednesday or so, basically any day where I have a lot of spare time to
troubleshoot any potential issues.

Let me know if you spot anything before said day.

**Pull request merged**